src/kbc_modules/cc_kbc/mod.rs

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) 2022 Alibaba Cloud
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use crate::kbc_modules::{KbcCheckInfo, KbcInterface};
+
+mod attester;
+mod crypto;
+mod kbs_protocol;
+
+use anyhow::*;
+use attester::{detect_tee_type, Attester, TEE};
+use crypto::{hasher, TeeKey};
+use kbs_protocol::message::*;
+
+pub struct Kbc {
+    tee: String,
+    token: Option<String>,
+    nonce: String,
+    tee_key: Option<TeeKey>,
+    attester: Option<Box<dyn Attester + Send + Sync>>,
+}
+
+impl KbcInterface for Kbc {
+    fn check(&self) -> Result<KbcCheckInfo> {
+        Err(anyhow!("Check API of this KBC is unimplement."))
+    }
+
+    fn decrypt_payload(&mut self, _annotation: &str) -> Result<Vec<u8>> {
+        Err(anyhow!("Decrypt Payload API of this KBC is unimplement."))
+    }
+}
+
+impl Kbc {
+    pub fn new(_kbs_uri: String) -> Kbc {
+        // Detect TEE type of the current platform.
+        let tee_type = detect_tee_type().ok().unwrap_or(TEE::Unknown);
+
+        // Create attester instance.
+        let attester = tee_type.to_attester().ok();
+
+        Kbc {
+            tee: tee_type.to_string(),
+            token: None,
+            nonce: String::default(),
+            tee_key: TeeKey::new().ok(),
+            attester,
+        }
+    }
+
+    fn generate_evidence(&self) -> Result<Evidence> {
+        let key = self
+            .tee_key
+            .as_ref()
+            .ok_or_else(|| anyhow!("Generate TEE key failed"))?;
+        let attester = self
+            .attester
+            .as_ref()
+            .ok_or_else(|| anyhow!("TEE attester missed"))?;
+
+        let tee_pubkey = key
+            .export_pubkey()
+            .map_err(|e| anyhow!("Export TEE pubkey failed: {:?}", e))?;
+        let tee_pubkey_string = serde_json::to_string(&tee_pubkey)?;
```

Comment:
Is the order of the fields in this JSON string identical to what is produced in line 81 (tee_pubkey as part of the Evidence)? Is it defined as a compact JSON string (no spaces, no end-of-line chars)?

We have this in the [evidence json](https://github.com/confidential-containers/kbs/blob/main/docs/kbs_attestation_protocol.md#attestation):

```
	    /* The public key generated by KBC in HW-TEE, it is valid until the next time an attestation is required.
	       Its hash needs to be included in HW-TEE evidence and signed by HW-TEE hardware. */
	    "tee-pubkey": {
		    "algorithm": "",
		    "pubkey-length": "",
		    "pubkey": ""
		},
```

The KBS should be able to serialize this object to compute the hash of nonce+tee_pubkey_string (to verify the report). And the serialization must be identical to how this was done here (in the KBC).

Reply:
Thanks for the reminder. It is really necessary to ensure that the JSON string format is strictly consistent. Otherwise, KBS will not be able to verify the hash correctly. Therefore, we'd better set the tee-pubkey (line 81) contained in evidence to the `String` type, that is, defined as a compact JSON string, right?

Reply:
Option 1 is to set use a string value for `tee-pubkey`, for example:

```
{
  "evidence": {
    "nonce": "",
    "tee": "",
    "tee-pubkey": "{\"algorithm\":\"rsa-pkcs1v15\",\"pubkey-length\":2048,\"pubkey\":\"AABBCCDD...\"}",
    "tee-evidence": {
      ..
    }
  }
}
```

Option 2 is to keep the current format, but the JSON serialization (before hashing) should be done by a canonical-json serializer (both in KBC and KBS). Unfortunately, I don't think serde_json can do this out of the box. I see there's a [cjson crate](https://docs.rs/cjson/latest/cjson/) and a [canonical_json crate](https://docs.rs/canonical_json/latest/canonical_json/) but both look a bit unmaintained. Also there's no clear standard for canonical JSON (or worse: [xkcd 927](https://xkcd.com/927/)).

Reply:
So, option 1 looks better now.